x-pack/plugins/transform/public/app/sections/transform_management/components/edit_transform_flyout/edit_transform_retention_policy.tsx

```diff
@@ -10,75 +10,113 @@ import React, { useEffect, useMemo, useState, type FC } from 'react';
-
-import { KBN_FIELD_TYPES } from '@kbn/field-types';
-
-import { useAppDependencies } from '../../../../app_dependencies';
+import { toMountPoint } from '@kbn/react-kibana-mount';
+import { useAppDependencies, useToastNotifications } from '../../../../app_dependencies';
+import { ToastNotificationText } from '../../../../components';
+import { isLatestTransform, isPivotTransform } from '../../../../../../common/types/transform';
+import { useGetTransformsPreview } from '../../../../hooks';
 
 import { EditTransformFlyoutFormTextInput } from './edit_transform_flyout_form_text_input';
 import { useEditTransformFlyout } from './use_edit_transform_flyout';
+import { getErrorMessage } from '../../../../../../common/utils/errors';
 
 export const EditTransformRetentionPolicy: FC = () => {
-  const appDeps = useAppDependencies();
-  const dataViewsClient = appDeps.data.dataViews;
+  const { i18n: i18nStart, theme } = useAppDependencies();
+
+  const toastNotifications = useToastNotifications();
 
   const dataViewId = useEditTransformFlyout('dataViewId');
   const formSections = useEditTransformFlyout('stateFormSection');
   const retentionPolicyField = useEditTransformFlyout('retentionPolicyField');
   const { formField, formSection } = useEditTransformFlyout('actions');
+  const requestConfig = useEditTransformFlyout('config');
 
-  const [dateFieldNames, setDateFieldNames] = useState<string[]>([]);
+  const previewRequest = useMemo(() => {
+    return {
+      source: requestConfig.source,
+      ...(isPivotTransform(requestConfig) ? { pivot: requestConfig.pivot } : {}),
+      ...(isLatestTransform(requestConfig) ? { latest: requestConfig.latest } : {}),
+    };
+  }, [requestConfig]);
+  const [destIndexAvailableTimeFields, setDestIndexAvailableTimeFields] = useState<string[]>([]);
+
+  const { error: transformsPreviewError, data: transformPreview } =
+    useGetTransformsPreview(previewRequest);
 
   useEffect(
     function getDateFields() {
       let unmounted = false;
-      if (dataViewId !== undefined) {
-        dataViewsClient.get(dataViewId).then((dataView) => {
-          if (dataView) {
-            const dateTimeFields = dataView.fields
-              .filter((f) => f.type === KBN_FIELD_TYPES.DATE)
-              .map((f) => f.name)
-              .sort();
-            if (!unmounted) {
-              setDateFieldNames(dateTimeFields);
-            }
+      if (transformPreview !== undefined) {
+        if (transformPreview.generated_dest_index) {
+          const properties = transformPreview.generated_dest_index.mappings.properties;
+          const timeFields: string[] = Object.keys(properties).filter(
+            (col) => properties[col].type === 'date'
+          );
+
+          if (!unmounted) {
+            setDestIndexAvailableTimeFields(timeFields);
           }
-        });
+        }
         return () => {
           unmounted = true;
         };
       }
     },
-    [dataViewId, dataViewsClient]
+    [transformPreview]
```

Comment:
There is no async code anymore, hence the entire effect and state can be replace with `useMemo`

Reply:
refactored in https://github.com/elastic/kibana/pull/172609/commits/40c4feb39c764fd023f19393b57eaca1615290db